components/molecules/MainSearchBar/MainSearchBar.tsx

```diff
@@ -1,44 +1,189 @@
-function TagCheckbox({ isTag }: IIsTag) {
+interface ITagSearchBarProps {
+  addTagList: (tag: ITag) => void;
+  wholeTagList: ITag[];
+  onClick: () => void;
+}
+
+function TagCheckbox({
+  onCheckboxClick,
+  isChecked,
+}: {
+  onCheckboxClick: () => void;
+  isChecked: boolean;
+}) {
   return (
     <S.TagValidation>
-      <input type="checkbox" />
+      <input type="checkbox" onClick={onCheckboxClick} checked={isChecked} />
       <div>태그 검색하기</div>
     </S.TagValidation>
   );
 }
 
-function SearchBar() {
+function TagSearchBar({
+  addTagList,
+  wholeTagList,
+  onClick,
+}: ITagSearchBarProps) {
+  const [inputValue, setInputValue] = useState('');
+  const input = useRef<HTMLInputElement>(null);
+
   const onSubmit = (e: React.FormEvent) => {
     e.preventDefault();
-    alert('searchbar를 클릭하셨군요!');
+  };
+  const onFilteredTagClick = (tag: ITag) => {
+    addTagList(tag);
+    setInputValue('');
+    if (input.current) {
+      input.current.value = '';
+    }
   };
 
   return (
-    <S.SearchBarForm onSubmit={onSubmit}>
-      <input type="text" />
+    <S.TagSearchBarContainer>
+      <S.SearchBarForm onSubmit={(e) => onSubmit(e)}>
+        <input
+          type="text"
+          placeholder="tagSearch"
+          onKeyUp={(e: React.KeyboardEvent<HTMLInputElement>) => {
+            setInputValue(e.target.value);
+          }}
+          ref={input}
+        />
+        <button type="button">
+          <Image src={searchIcon} alt="search icon" />
+        </button>
+      </S.SearchBarForm>
+      {inputValue !== '' ? (
+        <S.FilteredTagList>
+          {wholeTagList
+            .filter(({ name }) => name.includes(inputValue))
+            .map(({ id, name, color }) => (
+              <S.FilteredTag
+                key={id}
+                color={color}
+                onClick={() => {
+                  onFilteredTagClick({ id, name, color });
+                  onClick();
+                }}
+              >
+                {name}
+              </S.FilteredTag>
+            ))}
+        </S.FilteredTagList>
+      ) : (
+        <S.FilteredTagList>
+          {wholeTagList.map(({ id, name, color }) => (
+            <S.FilteredTag
+              key={id}
+              color={color}
+              onClick={() => {
+                onFilteredTagClick({ id, name, color });
+                onClick();
+              }}
+            >
+              {name}
+            </S.FilteredTag>
+          ))}
+        </S.FilteredTagList>
+      )}
+    </S.TagSearchBarContainer>
+  );
+}
+
+function TitleSearchBar({
+  handleKeyWord,
+}: {
+  handleKeyWord: (keyWord: string) => void;
+}) {
+  const input = useRef<HTMLInputElement>(null);
+  const onSubmit = (e: React.FormEvent) => {
+    e.preventDefault();
+    // 이 부분에서 도저히 any를 쓰지 않고 어떻게 해야할 지 잘 모르겠습니다 ㅜㅠ
+    const { target }: { target: any } = e;
+    handleKeyWord(target[0].value);
+  };
+  return (
+    <S.SearchBarForm onSubmit={(e) => onSubmit(e)}>
```

Comment:
제가 알기로는 이벤트 객체는 암묵적으로 전달되는 것으로 알고 있습니다.
```suggestion
    <S.SearchBarForm onSubmit={onSubmit}>
```